=== modified file 'wui/site/prado/Util/TLogRouter.php'
--- wui/site/prado/Util/TLogRouter.php	2010-07-20 07:43:16 +0000
+++ wui/site/prado/Util/TLogRouter.php	2011-05-18 10:38:50 +0000
@@ -830,7 +830,7 @@
 		$db=$this->getDbConnection();
 		$db->setActive(true);
 
-		$sql='SELECT * FROM '.$this->_logTable.' WHERE 0';
+		$sql='SELECT * FROM '.$this->_logTable.' WHERE FALSE';
 		try
 		{
 			$db->createCommand($sql)->query()->close();
@@ -838,10 +838,12 @@
 		catch(Exception $e)
 		{
 			// DB table not exists
+
 			if($this->_autoCreate)
 				$this->createDbTable();
 			else
-				throw new TConfigurationException('db_logtable_inexistent',$this->_logTable);
+                          //throw new TConfigurationException('db_logtable_inexistent',$this->_logTable);
+                          throw $e;
 		}
 
 		parent::init($config);

