=== modified file 'loudmouth/lm-connection.c'
--- loudmouth/lm-connection.c	2010-09-27 13:00:52 +0000
+++ loudmouth/lm-connection.c	2010-09-27 13:33:35 +0000
@@ -231,6 +231,13 @@
 	if (connection->auth_cb) {
 		_lm_utils_free_callback (connection->auth_cb);
 	}
+    if(connection->features_cb!=NULL)
+    {
+      lm_connection_unregister_message_handler(connection,
+                                               connection->features_cb,
+                                               LM_MESSAGE_TYPE_STREAM_FEATURES);
+      lm_message_handler_unref(connection->features_cb);
+    }
 
 	lm_connection_set_disconnect_function (connection, NULL, NULL, NULL);
 
@@ -1451,6 +1458,14 @@
 				      connection->server,
 				      connection_sasl_auth_finished);
 
+		if(connection->features_cb!=NULL)
+        {
+          lm_connection_unregister_message_handler(connection,
+                                                   connection->features_cb,
+                                                   LM_MESSAGE_TYPE_STREAM_FEATURES);
+
+          lm_message_handler_unref(connection->features_cb);
+        }
 		connection->features_cb  =
 			lm_message_handler_new (connection_features_cb,
 				NULL, NULL);

